django/core/files/utils.py

```diff
@@ -1,16 +1,23 @@
 import os
+import pathlib
 
 from django.core.exceptions import SuspiciousFileOperation
 
 
-def validate_file_name(name):
-    if name != os.path.basename(name):
-        raise SuspiciousFileOperation("File name '%s' includes path elements" % name)
-
+def validate_file_name(name, allow_relative_path=False):
     # Remove potentially dangerous names
-    if name in {'', '.', '..'}:
+    if os.path.basename(name) in {'', '.', '..'}:
         raise SuspiciousFileOperation("Could not derive file name from '%s'" % name)
 
+    if allow_relative_path:
+        path = pathlib.PurePosixPath(name)
```

Comment:
When we use `PurePosixPath` here we also should use that for `basename` to stay consistent, what do you think? I wonder if `PurePath` as opposed to the posix variant would be enough.

Reply:
I used `PurePosixPath()` because `filename` should be always a posix path in [`validate_file_name(filename, allow_relative_path=True)`](https://github.com/django/django/pull/14372/files/11f2850ab143a0ffe7d85be307d3da535522d7ab#diff-ffc2a7a9623222d5926482bec96348c039ad377a9f310da9c6ccdb3b7b17086dR321) call. Using `PurePath()` caused few failures on Windows, see [the previous build](https://djangoci.com/view/PR%20Builders/job/pull-requests-windows/11344/).

Reply:
@apollo13 We could add another kwarg `only_posix_path=False` and switch between `PurePath` and `PurePosixPath`. This would make this API clearer, IMO.

Reply:
Ok, so let's just ensure that we are using pathlib throughout the function then instead of a mix of `os.path` and `pathlib`

Reply:
OK, I added a comment. It's probably not worth extra complexity.